lib/membrane_mp4/muxer/cmaf/track_samples_cache.ex

```diff
@@ -0,0 +1,324 @@
+defmodule Membrane.MP4.Muxer.CMAF.TrackSamplesCache do
+  @moduledoc false
+
+  defstruct state: :collecting,
+            supports_keyframes?: false,
+            collected_duration: 0,
+            collected: [],
+            to_collect_duration: 0,
+            to_collect: []
+
+  @type state_t :: :collecting | :to_collect
+
+  @type t :: %__MODULE__{
+          state: state_t(),
+          supports_keyframes?: boolean(),
+          collected_duration: non_neg_integer(),
+          collected: list(Membrane.Buffer.t()),
+          to_collect: list(Membrane.Buffer.t())
+        }
+
+  @doc """
+  Force pushes the sample into the cache ignoring any timestamps.
+  """
+  @spec force_push(t(), Membrane.Buffer.t()) :: t()
+  def force_push(%__MODULE__{state: :collecting} = cache, sample) do
+    %__MODULE__{
+      cache
+      | collected: [sample | cache.collected],
+        collected_duration: cache.collected_duration + sample.metadata.duration
+    }
+  end
+
+  @spec push(t(), Membrane.Buffer.t(), Membrane.Time.t()) :: t()
+  def push(cache, sample, timestamp) do
```

Comment:
I think it's possible to reduce the number of variants of `push*` functions:
- Pushing audio segments is the same as pushing video segments if we mark each audio frame as a keyframe and set `min_timestamp` to equal `target_timestamp`
- Pushing non-partial segments is the same as pushing partial segments if we set `end_timestamp` to infinity

We may end up doing some redundant calculations, but I doubt that it will be a significant overhead. The size of this module should shrink significantly.

Reply:
> Pushing audio segments is the same as pushing video segments if we mark each audio frame as a keyframe and set min_timestamp to equal target_timestamp

To be honest I would avoid the approach of marking/marking all the audio samples as keyframes. That was the previous approach and it was a pure mess. I think that with this one it is probably better to separate them even if it produces more code.

> Pushing non-partial segments is the same as pushing partial segments if we set end_timestamp to infinity

Yup, you're right. How awesome that erlang's numbers are always smaller than atoms 😆